jcl/src/java.base/share/classes/openj9/internal/foreign/abi/UpcallMHMetaData.java

```diff
@@ -101,4 +108,93 @@ final class UpcallMHMetaData {
 		this.scope = ((scope != null) && (scope.ownerThread() != null)) ? scope : ResourceScope.newSharedScope();
 		/*[ENDIF] JAVA_SPEC_VERSION >= 20 */
 	}
+
+	/* Determine whether the memory segment of the passed-in/returned pointer is allocated
+	 * in the native memory or not and return its native address if valid.
+	 *
+	 * Note:
+	 * The method is shared in java (downcall) and in native (upcall) via the calling-in from the dispatcher.
+	 */
+	static long getNativeArgRetSegmentOfPtr(MemorySegment argRetSegmentOfPtr) {
+		/*[IF JAVA_SPEC_VERSION >= 20]*/
+		/* Verify MemorySegment.NULL as it is introduced since JDK20. */
+		if (argRetSegmentOfPtr == MemorySegment.NULL) {
+			throw new NullPointerException("A NULL memory segment is not allowed for pointer.");
+		}
+		/*[ENDIF] JAVA_SPEC_VERSION >= 20 */
+		if (!argRetSegmentOfPtr.isNative()) {
+			throw new IllegalArgumentException("Heap segment not allowed: " + argRetSegmentOfPtr);
+		}
+
+		/*[IF JAVA_SPEC_VERSION >= 20]*/
+		return argRetSegmentOfPtr.address();
+		/*[ELSE] JAVA_SPEC_VERSION >= 20 */
+		return argRetSegmentOfPtr.address().toRawLongValue();
+		/*[ENDIF] JAVA_SPEC_VERSION >= 20 */
+	}
+
+	/*[IF JAVA_SPEC_VERSION <= 19]*/
+	/* Determine whether the memory address of the passed-in/returned pointer is allocated
+	 * in the native memory or not and return its native address if valid.
+	 *
+	 * Note:
+	 * The method is shared in java (downcall) and in native (upcall) via the calling-in from the dispatcher.
+	 */
+	static long getNativeArgRetAddrOfPtr(MemoryAddress argRetAddrOfPtr) {
+		if (argRetAddrOfPtr == MemoryAddress.NULL) {
+			throw new NullPointerException("A NULL memory address is not allowed for pointer.");
+		}
+
+		/*[IF JAVA_SPEC_VERSION > 17]*/
+		/* Validate the native address as MemoryAddress.isNative() is removed in JDK18/19. */
+		if (argRetAddrOfPtr.toRawLongValue() == 0)
+		/*[ELSE] JAVA_SPEC_VERSION > 17 */
+		if (!argRetAddrOfPtr.isNative())
+		/*[ENDIF] JAVA_SPEC_VERSION > 17 */
+		{
+			throw new IllegalArgumentException("A heap address is not allowed: " + argRetAddrOfPtr);
+		}
+
+		return argRetAddrOfPtr.toRawLongValue();
+	}
+	/*[ENDIF] JAVA_SPEC_VERSION <= 19 */
```

Comment:
This could have been simpler: it only really applies to Java 17 (we don't need to worry about Java 18 or 19).

Reply:
The removal of the JDK18/19 related code will be included at https://github.com/eclipse-openj9/openj9/pull/17352.